Buttplug.Server/Bluetooth/Devices/MagicMotion.cs

```diff
@@ -4,127 +4,222 @@
+        }
+
+        private async Task<ButtplugMessage> HandleStopDeviceCmd(ButtplugDeviceMessage aMsg, CancellationToken aToken)
+        {
+            return await HandleSingleMotorVibrateCmd(new SingleMotorVibrateCmd(aMsg.DeviceIndex, 0, aMsg.Id), aToken);
+        }
+
+        private async Task<ButtplugMessage> HandleSingleMotorVibrateCmd(ButtplugDeviceMessage aMsg, CancellationToken aToken)
+        {
+            if (!(aMsg is SingleMotorVibrateCmd cmdMsg))
+            {
+                return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
+            }
+
+            return await HandleVibrateCmd(VibrateCmd.Create(cmdMsg.DeviceIndex, cmdMsg.Id, cmdMsg.Speed, _devInfo.VibeCount), aToken);
+        }
+
+        private async Task<ButtplugMessage> HandleVibrateCmd(ButtplugDeviceMessage aMsg, CancellationToken aToken)
+        {
+            if (!(aMsg is VibrateCmd cmdMsg))
+            {
+                return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
+            }
+
+            if (cmdMsg.Speeds.Count == 0 || cmdMsg.Speeds.Count > _devInfo.VibeCount )
+            {
+                return new Error(
+                    $"VibrateCmd requires between 1 and {_devInfo.VibeCount} vectors for this device.",
+                    Error.ErrorClass.ERROR_DEVICE,
+                    cmdMsg.Id);
+            }
+
+            var changed = false;
+
+            foreach (var v in cmdMsg.Speeds)
+            {
+                if (v.Index >= _devInfo.VibeCount)
+                {
+                    return new Error(
+                        $"Index {v.Index} is out of bounds for VibrateCmd for this device.",
+                        Error.ErrorClass.ERROR_DEVICE,
+                        cmdMsg.Id);
+                }
+
+                if (Math.Abs(v.Speed - _vibratorSpeeds[v.Index]) > 0.001)
+                {
+                    changed = true;
+                    _vibratorSpeeds[v.Index] = v.Speed;
+                }
+            }
+
+            if (!changed)
+            {
+                return new Ok(cmdMsg.Id);
+            }
+
+            byte[] data = null;
+            if (_devInfo.Protocol == 1)
+            {
+                data = new byte[] { 0x0b, 0xff, 0x04, 0x0a, 0x32, 0x32, 0x00, 0x04, 0x08, 0x00, 0x64, 0x00 };
+                data[9] = Convert.ToByte(_vibratorSpeeds[0] * byte.MaxValue);
+            }
+            else if (_devInfo.Protocol == 2)
+            {
+                data = new byte[] { 0x10, 0xff, 0x04, 0x0a, 0x32, 0x0a, 0x00, 0x04, 0x08, 0x00, 0x64, 0x00, 0x04, 0x08, 0x00, 0x64, 0x01 };
+                data[9] = Convert.ToByte(_vibratorSpeeds[0] * byte.MaxValue);
+                if (_devInfo.VibeCount >= 2)
+                {
+                    data[14] = Convert.ToByte(_vibratorSpeeds[1] * byte.MaxValue);
+                }
+            }
+
+            if (data == null)

```

Comment:
This could just be an else statement in the above logic block, though with an enum, you might be able to turn all of this into a switch and have this as the default block too.